### PR TITLE
Bento: unmount/remount component on DOM disconnect/reconnect

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #30952.

### CODE REVIEW

**Review by samouri — COMMENTED** [review]
Can you explain the difference between using `layoutCallback`/unlayoutCallback` vs. detached/attached cbs for this use case?  Doesn't layout imply attached and unlayout imply detached?

**Comment by dvoytenko** [comment]
@samouri 

> Can you explain the difference between using `layoutCallback`/`unlayoutCallback` vs. detached/attached cbs for this use case? Doesn't layout imply attached and unlayout imply detached?

There are two main concepts:

1. Rendering: whether the component is being rendered, callbacks are attached, listeners and observers setup. This has been historically in AMP driven by build callback. However, the main driver for this is whether the custom element is attached to DOM. For instance, the build scheduling is canceled if the element is removed from DOM.
2. Loading: whether the remote resource has been downloaded and decoded. This is done via poorly named `layoutCallback` and `unlayoutCallback`.

For (1) we've previously supported `buildCallback` and `detachedCallback` pair, b/c in AMP re-adding was not really a thing before - the document was sort of static and elements were one-off. In Bento we can't assume that. The `buildCallback` is a "first attachment callback" and it's not enough.

And one important note: (P)react in theory can separate rendering (DOM markup) and hydration (listeners, data structures, etc). But just as we've only supported it build as a one-off operation, Preact also supports it as a one-off operation. Because of this, I'm not unrendering the Preact component to free the resources. If in the future we get some sort of `unhydrate` API, then we can keep the DOM markup and rehydrate when a DOM node is reattached.

**Comment by samouri** [comment]
What causes a dom node to be detached and then reattached? Are there any actual situations where we'd want to leave the associated dom for the CE in place?

**Comment by dvoytenko** [comment]
> What causes a dom node to be detached and then reattached? Are there any actual situations where we'd want to leave the associated dom for the CE in place?

A simple user script would cause detach/attach. E.g.

```
newParent.appendChild(document.getElementById('timeago1'));
```

Being able to keep DOM intact is not a huge deal. But it'd (a) avoid initial FOUC on reappend, and (2) avoid two layouts.

**Comment by samouri** [comment]
> However, the main driver for this is whether the custom element is attached to DOM.

Does this mean you'd ideally want to phase `buildCallback` out and instead replace it with `attachedCallback`?

**Review by samouri — APPROVED** [review]

**Comment by dvoytenko** [comment]
> > However, the main driver for this is whether the custom element is attached to DOM.
> 
> Does this mean you'd ideally want to phase `buildCallback` out and instead replace it with `attachedCallback`?

Where I'd like to get is essentially almost the same API as custom elements which includes:
- constructor for one-time setup
- `connectedCallback` - for each time the element is attached.
- `disconnectedCallback` - for each time the element is detached.

It's still preferred to do one-time setup for many reasons. Everything we currently have in the `PreactBaseElement.buildCallback` would still be there. But we also definitely been overusing `buildCallback` for things that shouldn't be there in other components.

File: src/base-element.js [hunk 1]
```diff
@@ -334,6 +334,16 @@ export class BaseElement {
     // Subclasses may override.
   }
 
+  /**
+   * Override in subclass to adjust the element when it is being added to
+   * the DOM. Could e.g. be used to add a listener. Notice, that this
+   * callback is called immediately after `buildCallback()` if the element
+   * is attached to the DOM.
+   */
+  attachedCallback() {
```
Comment (caroqliu) [inline src/base-element.js:343]: This is great. @dmanek and I were just discussing in #30901 about where event listeners should be. It's really not consistent between components.

**Review by caroqliu — APPROVED** [review]

File: src/base-element.js (hunk 1 above)
Comment (dvoytenko) [reply src/base-element.js:343]: Yes. We definitely have an issue now. Hopefully all Bento components will only have external listeners running from within the Preact component itself. In that case it'd be handled fully automatically for us.

**Comment by dvoytenko** [comment]
@samouri by the way, with this PR we now have the mount/unmount pair of callbacks that corresponds to the `buildCallback`. So far we've avoided it for other work, but if we need it elsewhere, we can now use it. Thus:

1. Rendering: attach/detach
2. Loading: layout/unlayout

**Pull request merged**